opencensus/exporters/trace/ocagent/internal/ocagent_exporter.cc

```diff
@@ -0,0 +1,338 @@
+        static_cast<int32_t>(from_span.status().CanonicalCode()));
+    to_span->mutable_status()->set_message(from_span.status().error_message());
+
+    // 12. True if the parent is on a different process.
+    to_span->mutable_same_process_as_parent_span()->set_value(
+        !from_span.has_remote_parent());
+
+    // 13. child_span_count
+
+    // 14. span kind
+
+    // 15. tracestate
+
+    // 16. resource
+    // (TODO) Add FULL support.
+  }
+}
+
+Handler::Handler(OcAgentOptions &&opts) : opts_(std::move(opts)) {
+  InitNode();
+  ConnectAgent();
+}
+
+void Handler::Export(
+    const std::vector<::opencensus::trace::exporter::SpanData> &spans) {
+  ::opencensus::proto::agent::trace::v1::ExportTraceServiceRequest request;
+  ConvertSpans(spans, &request);
+  request.mutable_node()->CopyFrom(nodeInfo_);
+
+  grpc::ClientContext context;
+  context.set_deadline(absl::ToChronoTime(absl::Now() + opts_.rpc_deadline));
+
+  auto stream = opts_.trace_service_stub->Export(&context);
+  bool ok = stream->Write(request);
+}
+
+void Handler::InitNode() {
+  nodeInfo_.New();
+  auto identifier = nodeInfo_.mutable_identifier();
+
+  char host[100] = {0};
+  if (gethostname(host, sizeof(host)) < 0) {
+    snprintf(host, sizeof(host), "____default_host_name");
+  }
+  identifier->set_host_name(host);
+  identifier->set_pid(getpid());
+
+  {
+    auto ts_ptr = identifier->mutable_start_timestamp();
+    auto t = absl::Now();
+    const int64_t s = absl::ToUnixSeconds(t);
+    ts_ptr->set_seconds(s);
+    ts_ptr->set_nanos((t - absl::FromUnixSeconds(s)) / absl::Nanoseconds(1));
+  }
+
+  auto library_info = nodeInfo_.mutable_library_info();
+  library_info->set_language(
+      ::opencensus::proto::agent::common::v1::LibraryInfo_Language_CPP);
+  library_info->set_exporter_version("0.0.1");
+  library_info->set_core_library_version("0.0.1");
+
+  auto service_info = nodeInfo_.mutable_service_info();
+  service_info->set_name(host);
+}
+
+void Handler::ConnectAgent() {
+  ::opencensus::proto::agent::trace::v1::ExportTraceServiceRequest request;
+  request.mutable_node()->CopyFrom(nodeInfo_);
+
+  grpc::ClientContext context;
+  auto stream = opts_.trace_service_stub->Export(&context);
+
+  stream->Write(request);
+
+  ::opencensus::proto::agent::trace::v1::CurrentLibraryConfig cur_lib_cfg;
+  cur_lib_cfg.mutable_node()->CopyFrom(request.node());
+
+  auto config = cur_lib_cfg.mutable_config();
+  auto sampler = config->mutable_constant_sampler();
+  sampler->set_decision(::opencensus::proto::trace::v1::
+                            ConstantSampler_ConstantDecision_ALWAYS_ON);
+
+  grpc::ClientContext context1;
+  auto cfg_stream = opts_.trace_service_stub->Config(&context1);
+  cfg_stream->Write(cur_lib_cfg);
+}
+
+}  // namespace
+
+// static
+void OcAgentExporter::Register(OcAgentOptions &&opts) {
+  if (opts.trace_service_stub == nullptr) {
+    auto channel =
+        grpc::CreateChannel(opts.address, grpc::InsecureChannelCredentials());
```

Comment:
Should Insecure be `::grpc::GoogleDefaultCredentials()`?

Also please add ` ::opencensus::common::WithUserAgent()` for consistency with other gRPC exporters.

Reply:
OcAgent is always deploy in the same Pod with app, so I used InsecureChannelCredentials.
Can I add an option in `opts` as a label indicating witch kind of Channel Credential they want to use and keep Insecure as default?

Reply:
I think you're right and we should leave it as Insecure*

If users need a different setup, they can already supply their own stub via options.

IMO let's not add an option for channel credentials in this PR.

Resolving this.